### PR TITLE
Test all classes

### PR DESCRIPTION
**Pull request by jaferIdrees**
### Added Unittest to all methods in : 
- [ ] Book
- [ ] Person
- [ ] Student
- [ ] Teacher
- [ ] Classroom
- [ ] Rental
- [ ] Decorators
- [ ] CapitalizeDecorator
- [ ] TrimmerDecorator

### CODE REVIEW

File: spec/test_books_spec.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+require_relative '../classes/book'
+
+describe Book do
+  context 'when testing the Book class' do
+    it 'Create Book' do
+      book = Book.new('Oliver Twist', 'Jafer Yousef')
+
+      expect(book.author).to eq 'Jafer Yousef'
+      expect(book.title).to eq 'Oliver Twist'
+    end
```
Comment (BenMukebo) [inline spec/test_books_spec.rb]: - [ ] Great job so far 👏🏻 👏🏻 However, according to the project requirements, there should be unit tests for all the methods in all the classes, Kindly make sure the `add_rental` classbook method has (at least) an 'it' block 👇🏻 check your book.rb file

```ruby
  def add_rental(rental)
    @rentals.push(rental)
    rental.book = self
  end
```

File: spec/test_classroom_spec.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+describe ClassRoom do
+  context 'When testing the Classroom class' do
+    it 'should create classroom' do
+      student_one = Student.new(classroom: 'Maths class', age: '22', id: '222', name: 'Jafer Yousef')
+      student_two = Student.new(classroom: 'English class', age: '12', id: '122', name: 'Elias Yousef')
+
+      classroom = ClassRoom.new('Maths Class')
+      classroom.add_student(student_one)
+      classroom.add_student(student_two)
```
Comment (BenMukebo) [inline spec/test_classroom_spec.rb]: - [ ] Great job so far 👏🏻 👏🏻 However, according to the project requirements, there should be an 'it' block for each class method (at least). Kindly make sure the `add_student` class classroom method has (at least) an 'it' block

File: spec/test_person_spec.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+require_relative '../classes/person'
+
+describe Person do
+  context 'When creating a Person' do
+    it 'Should create a new Person' do
+      person = Person.new(34, 'Idrees')
+
+      expect(person.name).to eq 'Idrees'
+      expect(person.age).to eq 34
+    end
```
Comment (BenMukebo) [inline spec/test_person_spec.rb]: - [ ] Great job so far 👏🏻 👏🏻 However, according to the project requirements, there should be unit tests for all the methods in all the classes, Kindly make sure the `of_age?, can_use_services?,  and  correct_name` class person method has (at least) an 'it' block 👇🏻 check your person.rb file

```ruby
  def can_use_services?
    of_age? || @parent_permission
  end

  def correct_name
    @name
  end

  private

  def of_age?
    @age >= 18
  end
```

File: spec/test_teacher_spec.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+      expect(teacher.name).to eq 'Elias'
+    end
+
+    it "Should return '22' as a teacher age" do
+      expect(teacher.age).to eq 22
+    end
+
+    it "Should return 'Math' as a teacher specialization" do
+      expect(teacher.specialization).to eq 'Math'
+    end
```
Comment (BenMukebo) [inline spec/test_teacher_spec.rb:18]: - [x] Great job so far 👏🏻 👏🏻 🥳  Just one thing miss, according to the project requirements, there should be unit tests for all the methods in all the classes, Kindly make sure the `can_use_services?` class method has (at least) an 'it' block 👇🏻 check your teacher,rb file

```ruby
  # Method can_use_services?
  def can_use_services?
    true
  end
```

File: classes/teacher.rb [hunk 5]
```diff
@@ -3,7 +3,7 @@
 class Teacher < Person
   attr_accessor :specialization
 
-  def initialize(specialization, name, age, id:nil)
+  def initialize(specialization, name, age, id: nil)
     super(name, age, id: id)
```
Comment (BenMukebo) [inline classes/teacher.rb:7]: - [ ] There should be unit tests for the student classes, Kindly make sure to test all methods of the student class !!👇🏻 check your student,rb file


```ruby
 # Method play_hooky
  def play_hooky
    "¯\(ツ)/¯"
  end

**Review by BenMukebo — CHANGES_REQUESTED** [review]
### Hi @jaferIdrees, @Fourteen98,

You did a great job so far 👏🏼 👏🏼
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

### To highlight 💪🏻

- No linter errors. ✔️ 
- Gitflow was correctly used.  ✔️   
- There is one file per class. ✔️ 
 👍  Good job!


## STATUS: CHANGES REQUESTED  ♻️ 
_Please Check the comments under the review and make all required changes_

- [ ] Great job so far 👏🏻 👏🏻 However, according to the project requirements, there should be unit tests for all the methods in all the classes, Kindly make sure to add unit tests for all the methods of the: `Student class` ✍🏻 


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏 💻  
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me @BenMukebo when you comment so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/test_books_spec.rb (hunk 1 above)
Comment (jaferIdrees) [reply spec/test_books_spec.rb]: Hello  @BenMukebo
I believe this is no realistic the project requirements says to add tests for all classes not for all methods in all classes!

**Review by Lameck1 — COMMENTED** [review]
Hello Team 👋,

_This is an additional review due to the second opinion request._

Please let's all be informed that unit testing is a way of testing a unit - the smallest piece of code that can be logically isolated in a system. This means that all the functionality of the application should be tested. Let's not choose which method we are going to test and the ones we will not, that's not the goal of unit testing.

Now about your comment, "**_the project requirement says to add tests to all classes only_**", this is not true. We are not adding tests to the classes. This is the project requirement 👇🏼:

  - Add unit tests to your classes:
  
    The concern is `"Add unit tests to`", not `"add tests to all classes"`. The units are the functions you have in the class body. Therefore, there is a need for you to add tests for those functions(the units).

Please work on the suggested changes and request another code when done.

All the best.

Keep rocking 🚀!

**Review by ShadyShawkat — APPROVED** [review]
## Hi Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight
- All linters are passing ✔️ 
- All tests are passing ✔️ 
- Descriptive commit messages ✔️ 
- Professional looking pull request ✔️ 
- Your readme looks professional ✔️ 

Image

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**